### PR TITLE
core: fix file_get_by_tag()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Fixes a race in file_get_by_tag() where a file is just about to be freed
but hasn't been removed from the linked list yet.

Fixes: bae0f170e321 ("core: file: refactor interface to be thread safe")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: core/mm/file.c [hunk 1]
```diff
@@ -122,9 +122,32 @@ struct file *file_get_by_tag(const uint8_t *tag, unsigned int taglen)
...
+	 * If file is found and reference counter can be increased, we're done.
+	 * If file can't be found, it doesn't exist so it has to be added.
+	 * If it's found but reference counter is 0, the situation is
+	 * a bit complicated:
+	 * - file_put() is about to free the file as soon as it can obtain the
+	 *   mutex.
+	 * - Unless there's a mismatch between file_get() and file_put() only
+	 *   one thread calling file_put() is about to free the file.
+	 *
+	 * There's a window of oportunity where file_put() is called
```
Comment (jforissier) [inline core/mm/file.c]: opportunity

File: core/mm/file.c [hunk 2]
```diff
@@ -122,9 +122,32 @@ struct file *file_get_by_tag(const uint8_t *tag, unsigned int taglen)
...
+	 * (without a mutex being held, which is quite OK) while we're
+	 * holding the mutex here and are searching for the file and it's
+	 * found, but just after file_put() has decreased the reference
+	 * counter.
```
Comment (jforissier) [inline core/mm/file.c]: Wouldn't it be much simpler to always grab the mutex before touching anything?...

Comment (jenswi-linaro) [reply core/mm/file.c]: Yes it would, but then we'd have more congestion too.

Comment (jforissier) [reply core/mm/file.c]: Maybe, but I'd be curious about the performance impact. Anyway, this looks OK, although I'm still trying to convince myself there can be no other race condition...

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/mm/file.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/mm/file.c]: I have no numbers on performance impact.

**Comment by jenswi-linaro** [comment]
Comment addressed, tag applied

**Pull request merged**